File: 01-client-setup/steps/create-apns-key.js
const { execSync, spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const chalk = require('chalk');

/**
 * APNs Key Creator
 *
 * Creates APNs Authentication Keys (.p8) for iOS push notifications
 * using the Apple Developer Portal API via Ruby/Spaceship.
 *
 * NOTE: This requires interactive authentication (Apple ID + 2FA)
 * The operator will be prompted for password and verification code.
 */
class APNsKeyCreator {
  constructor() {
    this.scriptPath = path.resolve(__dirname, '../../02-build-deploy/scripts/create_apns_key.rb');
    this.wrapperPath = path.resolve(__dirname, '../../02-build-deploy/scripts/run_apns_script.sh');
    // loyalty-credentials is a sibling repo to loyalty-compose, not inside it
    // From: 01-client-setup/steps/ -> ../../../loyalty-credentials
    this.credentialsPath = path.resolve(__dirname, '../../../loyalty-credentials');
    this.apnsDir = path.join(this.credentialsPath, 'shared', 'apns');
  }

  /**
   * Check if an APNs key already exists
   * @returns {Object} { exists: boolean, keyFile?: string, keyId?: string }
   */
  checkExistingKey() {
    if (!fs.existsSync(this.apnsDir)) {
      return { exists: false, reason: 'APNs directory not found' };
    }

    const files = fs.readdirSync(this.apnsDir);
    const p8Files = files.filter((f) => f.endsWith('.p8'));

    if (p8Files.length === 0) {
      return { exists: false, reason: 'No .p8 files found' };
    }

    // Extract Key ID from filename (AuthKey_XXXXXXXX.p8)
    const keyFile = p8Files[0];
    const keyIdMatch = keyFile.match(/AuthKey_([A-Z0-9]+)\.p8/);
    const keyId = keyIdMatch ? keyIdMatch[1] : null;

    return {
      exists: true,
      keyFile: path.join(this.apnsDir, keyFile),
      keyId,
      allKeys: p8Files,
    };
  }

  /**
   * Get Team ID from environment or credentials
   * @returns {string|null} Team ID
   */
  getTeamId() {
    return process.env.APPLE_TEAM_ID || '84LT77P2DM';
  }

  /**
   * Create APNs key interactively
   * @param {Object} options Configuration options
   * @param {Object} options.logger Logger instance
   * @param {Function} options.inquirer Inquirer instance for prompts
   * @returns {Promise<Object>} Result with keyId, teamId, keyPath
   */
  async createKey(options = {}) {
    const { logger, inquirer } = options;

    // Check if key already exists
    const existingKey = this.checkExistingKey();
    if (existingKey.exists) {
      if (logger) {
        logger.info(`✅ APNs key already exists: ${existingKey.keyFile}`);
        logger.info(`   Key ID: ${existingKey.keyId}`);
      }
      return {
        success: true,
        skipped: true,
        reason: 'Key already exists',
        keyId: existingKey.keyId,
        keyFile: existingKey.keyFile,
        teamId: this.getTeamId(),
      };
    }

    // Check if Apple Developer Email is configured
    const appleEmail = process.env.APPLE_DEVELOPER_EMAIL;
    if (!appleEmail) {
      if (logger) {
        logger.error('APPLE_DEVELOPER_EMAIL environment variable is required');
        logger.info('Add it to automation/.env file');
      }
      return {
        success: false,
        error: 'APPLE_DEVELOPER_EMAIL not configured',
      };
    }

    // Check if script exists
    if (!fs.existsSync(this.wrapperPath)) {
      if (logger) {
        logger.error(`APNs creation script not found: ${this.wrapperPath}`);
      }
      return {
        success: false,
        error: 'Script not found',
      };
    }

    // Prompt user before starting (requires interaction)
    if (inquirer && logger) {
      logger.blank();
      logger.warn('⚠️  APNs Key Creation requires Apple ID authentication');
      logger.info(`   Email: ${appleEmail}`);
      logger.info('   You will be prompted for password and 2FA code');
      logger.blank();

      const { proceed } = await inquirer.prompt([
        {
          type: 'confirm',
          name: 'proceed',
          message: 'Proceed with APNs key creation?',
          default: true,
        },
      ]);

      if (!proceed) {
        logger.info('APNs key creation skipped by user');
        return {
          success: true,
          skipped: true,
          reason: 'Skipped by user',
        };
      }
    }

    // Execute the script
    if (logger) {
      logger.info('Creating APNs key...');
      logger.info('Please follow the prompts for Apple ID authentication');
      logger.blank();
    }

    return new Promise((resolve) => {
      const child = spawn('bash', [this.wrapperPath], {
        env: {
          ...process.env,
          APPLE_DEVELOPER_EMAIL: appleEmail,
          APNS_OUTPUT_DIR: this.apnsDir,
        },
        stdio: 'inherit', // Allow interactive input/output
        cwd: path.dirname(this.wrapperPath),
      });

      child.on('close', (code) => {
        if (code === 0) {
          // Check for created key
          const newKey = this.checkExistingKey();
          if (newKey.exists) {
            if (logger) {
              logger.blank();
              logger.success('✅ APNs key created successfully!');
              logger.info(`   Key ID: ${newKey.keyId}`);
              logger.info(`   File: ${newKey.keyFile}`);
            }
            resolve({
              success: true,
              keyId: newKey.keyId,
              keyFile: newKey.keyFile,
              teamId: this.getTeamId(),
            });
          } else {
            resolve({
              success: false,
              error: 'Key file not found after creation',
            });
          }
        } else {
          if (logger) {
            logger.error(`APNs key creation failed with exit code: ${code}`);
          }
          resolve({
            success: false,
            error: `Script exited with code ${code}`,
          });
        }
      });

      child.on('error', (err) => {
        if (logger) {
          logger.error(`Failed to execute APNs script: ${err.message}`);
        }
        resolve({
          success: false,
          error: err.message,
        });
      });
    });
  }

  /**
   * Get APNs key info for Firebase upload instructions
   * @returns {Object|null} Key info or null if not found
   */
  getKeyInfo() {
    const existingKey = this.checkExistingKey();
    if (!existingKey.exists) {
      return null;
    }

    return {
      keyId: existingKey.keyId,
      keyFile: existingKey.keyFile,
      teamId: this.getTeamId(),
    };
  }
}

module.exports = APNsKeyCreator;

File: 02-build-deploy/scripts/create_apns_key.rb
#!/usr/local/opt/ruby/bin/ruby
# frozen_string_literal: true

# Script to create APNs Authentication Key (.p8) for push notifications
# This script includes a workaround for the 'scope' parameter issue
# See: https://github.com/fastlane/fastlane/issues/29450
#
# USAGE:
#   APPLE_DEVELOPER_EMAIL=your@email.com ./scripts/create_apns_key.rb
#
# Or with wrapper script:
#   ./scripts/run_apns_script.sh

# Setup gem path to use fastlane's gems
ENV['GEM_HOME'] ||= "#{ENV['HOME']}/.local/share/fastlane/3.4.0"
ENV['GEM_PATH'] = "#{ENV['GEM_HOME']}:/usr/local/Cellar/fastlane/2.229.0/libexec"

require 'spaceship'
require 'fileutils'
require 'json'

# Configuration
KEY_NAME = ENV['APNS_KEY_NAME'] || 'LoyaltyHub Push Key'
# loyalty-credentials is a sibling repo to loyalty-compose, not inside it
# From: 02-build-deploy/scripts/ -> ../../../loyalty-credentials/shared/apns
OUTPUT_DIR = ENV['APNS_OUTPUT_DIR'] || File.expand_path('../../../loyalty-credentials/shared/apns', __dir__)
USERNAME = ENV['APPLE_DEVELOPER_EMAIL']

unless USERNAME
  puts "❌ Error: APPLE_DEVELOPER_EMAIL environment variable is required"
  puts "   Usage: APPLE_DEVELOPER_EMAIL=your@email.com ruby #{$PROGRAM_NAME}"
  exit 1
end

puts "🔔 Creating APNs Authentication Key..."
puts "   Key Name: #{KEY_NAME}"
puts "   Output: #{OUTPUT_DIR}"
puts ""
puts "⚠️  NOTE: APNs Key management requires Apple ID authentication"
puts "   You will be prompted for password and 2FA code"
puts ""

# Login to Apple Developer Portal
puts "   Authenticating as: #{USERNAME}"
Spaceship::Portal.login(USERNAME)

# Get Team ID
team_id = Spaceship::Portal.client.team_id
puts "   Team ID: #{team_id}"

# Check existing keys
puts "   Checking existing APNs keys..."
all_keys = Spaceship::Portal::Key.all
apns_keys = all_keys.select { |k| (k.services rescue []).include?('APNS') }

puts "   Found #{all_keys.length} total key(s), #{apns_keys.length} APNs key(s)"

if apns_keys.length >= 2
  puts ""
  puts "⚠️  You already have #{apns_keys.length} APNs keys (Apple limit is 2)"
  apns_keys.each do |k|
    puts "   - #{k.name} (ID: #{k.id})"
  end
  puts ""
  puts "   To create a new key, you must revoke an existing one first."
  puts "   Use: fastlane revoke_apns_key key_id:XXXXXXXXXX"
  exit 1
end

# Create new APNs key with scope fix
puts ""
puts "   Creating new APNs key: #{KEY_NAME}..."

begin
  # Apple Service IDs (from PR #29458)
  APNS_ID = 'U27F4V844T'

  # Build service config with v2 API format (fix from PR #29458)
  # The v2 API uses serviceConfigurationsRequests array with isNew flag
  service_configs_requests = [
    {
      'isNew' => true,
      'serviceId' => APNS_ID,
      'identifiers' => {},
      'environment' => 'all',
      'scope' => 'team'
    }
  ]

  # Create key using v2 API endpoint directly (bypassing broken spaceship method)
  client = Spaceship::Portal.client

  # Fetch CSRF token first (required for key operations)
  client.send(:fetch_csrf_token_for_keys)

  params = {
    name: KEY_NAME,
    serviceConfigurationsRequests: service_configs_requests,
    teamId: client.team_id
  }

  # Use v2 endpoint
  response = client.send(:request, :post, 'account/auth/key/v2/create') do |req|
    req.headers['Content-Type'] = 'application/json'
    req.body = params.to_json
  end

  parsed = client.send(:parse_response, response, 'keys').first
  new_key = Spaceship::Portal::Key.new(parsed)

  # Download the key (can only be done once!)
  key_content = new_key.download
  key_id = new_key.id

  # Ensure output directory exists
  FileUtils.mkdir_p(OUTPUT_DIR)

  # Save the .p8 file
  key_filename = "AuthKey_#{key_id}.p8"
  key_path = File.join(OUTPUT_DIR, key_filename)
  File.write(key_path, key_content)

  # Update APNS_KEY_INFO.md if it exists
  info_path = File.join(OUTPUT_DIR, 'APNS_KEY_INFO.md')
  if File.exist?(info_path)
    info_content = File.read(info_path)
    info_content.gsub!(/\| \*\*Key File\*\* \| `[^`]*` \|/, "| **Key File** | `#{key_filename}` |")
    info_content.gsub!(/\| \*\*Key ID\*\* \| `[^`]*` \|/, "| **Key ID** | `#{key_id}` |")
    info_content.gsub!(/\| \*\*Team ID\*\* \| `[^`]*` \|/, "| **Team ID** | `#{team_id}` |")
    info_content.gsub!(/\| \*\*Created\*\* \| [^\|]* \|/, "| **Created** | #{Time.now.strftime('%Y-%m-%d')} |")
    info_content.gsub!(/\| \*\*Uploaded to Firebase\*\* \| \[.\] [^\|]* \|/, "| **Uploaded to Firebase** | [ ] Not yet |")
    File.write(info_path, info_content)
    puts "   ✓ Updated APNS_KEY_INFO.md"
  end

  puts ""
  puts "✅ APNs Key created successfully!"
  puts ""
  puts "   📁 Key File: #{key_path}"
  puts "   🔑 Key ID: #{key_id}"
  puts "   👥 Team ID: #{team_id}"
  puts ""
  puts "⚠️  IMPORTANT: This key can only be downloaded ONCE!"
  puts "   Make sure to backup the .p8 file securely."
  puts ""
  puts "📋 Next step: Upload to Firebase Console"
  puts "   https://console.firebase.google.com/project/loyalty-hub-1f47c/settings/cloudmessaging"
  puts ""
  puts "   When uploading, use:"
  puts "   - Key ID: #{key_id}"
  puts "   - Team ID: #{team_id}"

rescue => e
  puts ""
  puts "❌ Failed to create APNs key: #{e.message}"

  if e.message.include?("maximum number")
    puts "   Apple allows maximum 2 APNs keys per account"
    puts "   Revoke an existing key first"
  end

  puts ""
  puts "Debug info:"
  puts e.backtrace.first(5).join("\n")
  exit 1
end
